crates/swc_ecma_transforms_compat/src/es2015/object_super.rs

```diff
@@ -0,0 +1,623 @@
+use std::iter;
+
+use swc_common::{util::take::Take, Span, DUMMY_SP};
+use swc_ecma_ast::*;
+use swc_ecma_transforms_base::helper;
+use swc_ecma_utils::{
+    alias_ident_for, is_rest_arguments, prepend, private_ident, quote_ident, ExprFactory, IdentExt,
+};
+use swc_ecma_visit::{as_folder, noop_visit_mut_type, Fold, VisitMut, VisitMutWith};
+
+struct ObjectSuper {
+    extra_var: Vec<Ident>,
+}
+pub fn object_super() -> impl Fold + VisitMut {
+    as_folder(ObjectSuper {
+        extra_var: Vec::new(),
+    })
+}
+
+impl VisitMut for ObjectSuper {
+    noop_visit_mut_type!();
+    fn visit_mut_module_items(&mut self, n: &mut Vec<ModuleItem>) {
+        n.visit_mut_children_with(self);
+        if self.extra_var.len() != 0 {
+            prepend(
+                n,
+                ModuleItem::Stmt(Stmt::Decl(Decl::Var(VarDecl {
+                    span: DUMMY_SP,
+                    kind: VarDeclKind::Var,
+                    declare: false,
+                    decls: self
+                        .extra_var
+                        .take()
+                        .into_iter()
+                        .map(|v| VarDeclarator {
+                            span: DUMMY_SP,
+                            name: Pat::Ident(v.into()),
+                            init: None,
+                            definite: false,
+                        })
+                        .collect(),
+                }))),
+            );
+        }
+    }
+    fn visit_mut_block_stmt(&mut self, stmts: &mut BlockStmt) {
+        stmts.visit_mut_children_with(self);
```

Comment:
I found a problem.
This will take all variables even it should not.


```js

const foo = {
    foo() {
        super.bar();
        const nested = {
            nesteedMethod() {
                super.foo();
            }
        };
    }
};


```


The variable for `super.bar();` will be injected into `nesteedMethod`.

To prevent this, you can preserve original variables like

```rust
fn visit_mut_block_stmt(&mut self, s: &mut BlockStmt) {
    let orig_vars = self.extra_var.take();

    s.visit_mut_children_with(self);

    // Prepend

    self.extra_var = orig_vars;
}

```

Reply:
I think `self.extra_var` is empty before `visit_mut_children_with(self)`.
since `self .extra_vars.drain(..)`